pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+				weight += process_runtime_upgrades::<T>();
+			}
+
+			weight
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_fully_upgraded)]
+	/// True if all required migrations have completed
+	type FullyUpgraded<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn migration_state)]
+	/// MigrationState tracks the progress of a migration.
+	/// Maps name (Vec<u8>) -> migration progress (Perbill)
+	type MigrationState<T: Config> = StorageMap<_, Twox64Concat, Vec<u8>, Perbill, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub completed_migrations: Vec<Vec<u8>>,
+		pub dummy: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				completed_migrations: vec![],
+				dummy: PhantomData,
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for migration_name in &self.completed_migrations {
+				<MigrationState<T>>::insert(migration_name, Perbill::one());
+			}
+		}
+	}
+
+	fn process_runtime_upgrades<T: Config>() -> Weight {
+		log::info!("stepping runtime upgrade");
+
+		// TODO: query proper value or make configurable
+		let available_weight: Weight = 500_000_000_000u64.into();
+		let mut weight: Weight = 0u64.into();
+		let mut done: bool = true;
+
+		for migration in &T::MigrationsList::get() {
+			// let migration_name = migration.friendly_name();
+			let migration_name = migration.friendly_name();
+			let migration_name_as_bytes = migration_name.as_bytes();
+			log::trace!("evaluating migration {}", migration_name);
+
+			let migration_state =
+				<MigrationState<T>>::get(migration_name_as_bytes).unwrap_or(Perbill::zero());
+
+			if migration_state < Perbill::one() {
+				if migration_state.is_zero() {
+					<Pallet<T>>::deposit_event(Event::MigrationStarted(
+						migration_name_as_bytes.into(),
+					));
+				}
+
+				let available_for_step = available_weight - weight;
+				log::trace!(
+					"stepping migration {}, prev: {:?}, avail weight: {}",
+					migration_name,
+					migration_state,
+					available_for_step
+				);
+
+				// perform a step of this migration
+				let (updated_progress, consumed_weight) =
+					migration.step(migration_state, available_for_step);
+				// TODO: error if progress == 0 still?
+				<Pallet<T>>::deposit_event(Event::MigrationStepped(
+					migration_name_as_bytes.into(),
+					updated_progress,
+					consumed_weight,
+				));
+
+				weight += consumed_weight;
+				if weight > available_weight {
+					// TODO: the intent here is to complain obnoxiously so that this is caught
+					// during development. In production, this should probably be tolerated because
+					// failing is catastrophic.
+					log::error!(
+						"Migration {} consumed more weight than it was given! ({} > {})",
+						migration_name,
+						consumed_weight,
+						available_for_step
+					);
+				}
```

Comment:
I wonder if `available_weight` should be set at some fraction of the total available weight so that `weight` can go above `available_weight` (to a limited extent) without bricking the chain.

Reply:
Eh never mind, not a panic and might work anyway if time is < 6 seconds